libkbfs/file_data.go

```diff
@@ -509,44 +535,298 @@ func (fd *fileData) createIndirectBlock(
+// returns the new top block (whether or not it is new), and also
+// returns any newly-dirtied block pointers.
 func (fd *fileData) newRightBlock(
-	ctx context.Context, ptr BlockPointer, pblock *FileBlock,
-	off int64) error {
-	newRID, err := fd.crypto.MakeTemporaryBlockID()
-	if err != nil {
-		return err
+	ctx context.Context, parentBlocks []parentBlockAndChildIndex, off int64,
+	df *dirtyFile, dver DataVer) (*FileBlock, []BlockPointer, error) {
+	// Find the lowest block that can accommodate a new right block.
+	lowestAncestorWithRoom := -1
+	for i := len(parentBlocks) - 1; i >= 0; i-- {
+		pb := parentBlocks[i]
+		if len(pb.pblock.IPtrs) < fd.bsplit.MaxPtrsPerBlock() {
+			lowestAncestorWithRoom = i
+			break
+		}
 	}
-	rblock := &FileBlock{}
 
-	newPtr := BlockPointer{
-		ID:      newRID,
-		KeyGen:  fd.kmd.LatestKeyGeneration(),
-		DataVer: DefaultNewBlockDataVersion(false),
-		Context: kbfsblock.Context{
-			Creator:  fd.uid,
-			RefNonce: kbfsblock.ZeroRefNonce,
-		},
+	var newTopBlock *FileBlock
+	var newDirtyPtrs []BlockPointer
+	if lowestAncestorWithRoom < 0 {
+		// Create a new level of indirection at the top.
+		var err error
+		newTopBlock, err = fd.createIndirectBlock(ctx, df, dver)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// The old top block needs to be cached under its new ID if it
+		// was indirect.
+		if len(parentBlocks) > 0 {
+			ptr := newTopBlock.IPtrs[0].BlockPointer
+			err = fd.cacher(ptr, parentBlocks[0].pblock)
+			if err != nil {
+				return nil, nil, err
+			}
+			newDirtyPtrs = append(newDirtyPtrs, ptr)
+		}
+
+		parentBlocks = append([]parentBlockAndChildIndex{{newTopBlock, 0}},
+			parentBlocks...)
+		lowestAncestorWithRoom = 0
 	}
 
-	pblock.IPtrs = append(pblock.IPtrs, IndirectFilePtr{
-		BlockInfo: BlockInfo{
-			BlockPointer: newPtr,
-			EncodedSize:  0,
-		},
-		Off: off,
-	})
+	fd.log.CDebugf(ctx, "Making new right block at off %d for file %v, "+
+		"lowestAncestor at level %d", off, fd.rootBlockPointer(),
+		lowestAncestorWithRoom)
+
+	// Make a new right block for every parent, starting with the
+	// lowest ancestor with room.  Note that we're not iterating over
+	// the actual parent blocks here; we're only using its length to
+	// figure out how many levels need new blocks.
+	pblock := parentBlocks[lowestAncestorWithRoom].pblock
+	for i := lowestAncestorWithRoom; i < len(parentBlocks); i++ {
+		newRID, err := fd.crypto.MakeTemporaryBlockID()
+		if err != nil {
+			return nil, nil, err
+		}
 
-	err = fd.cacher(newPtr, rblock)
-	if err != nil {
-		return err
+		newPtr := BlockPointer{
+			ID:      newRID,
+			KeyGen:  fd.kmd.LatestKeyGeneration(),
+			DataVer: dver,
+			Context: kbfsblock.Context{
+				Creator:  fd.uid,
+				RefNonce: kbfsblock.ZeroRefNonce,
+			},
+		}
+
+		fd.log.CDebugf(ctx, "New right block for file %d, level %d, ptr %v",
+			fd.rootBlockPointer(), i, newPtr)
+
+		pblock.IPtrs = append(pblock.IPtrs, IndirectFilePtr{
```

Comment:
There's another aliasing issue here for the `i = lowestAncestorWithRoom` iteration, since that comes from `parentBlocks`, which are blocks-for-reading, right?